zebra-network/src/meta_addr.rs

```diff
@@ -15,17 +15,112 @@ use zebra_chain::serialization::{
 
 use crate::protocol::types::PeerServices;
 
+/// Peer connection state, based on our interactions with the peer.
+///
+/// Zebra also tracks how recently a peer has sent us messages, and derives peer
+/// liveness based on the current time. This derived state is tracked using
+/// [`AddressBook::maybe_connected_peers`] and
+/// [`AddressBook::reconnection_peers`].
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum PeerConnectionState {
+    /// The peer has sent us a valid message.
+    ///
+    /// Peers remain in this state, even if they stop responding to requests.
+    /// (Peer liveness is derived from the `last_seen` timestamp, and the current
+    /// time.)
+    Responded,
+
+    /// The peer's address has just been fetched from a DNS seeder, or via peer
+    /// gossip, but we haven't attempted to connect to it yet.
+    NeverAttempted,
+
+    /// The peer's TCP connection failed, or the peer sent us an unexpected
+    /// Zcash protocol message, so we failed the connection.
+    Failed,
+
+    /// We just started a connection attempt to this peer.
+    AttemptPending,
+}
+
+impl Default for PeerConnectionState {
+    fn default() -> Self {
+        PeerConnectionState::NeverAttempted
+    }
+}
+
+impl Ord for PeerConnectionState {
+    /// `PeerConnectionState`s are sorted in approximate reconnection attempt
+    /// order, ignoring liveness.
+    ///
+    /// See [`CandidateSet`] and [`MetaAddr::cmp`] for more details.
+    fn cmp(&self, other: &Self) -> Ordering {
+        use PeerConnectionState::*;
+        match (self, other) {
+            (Responded, Responded)
+            | (NeverAttempted, NeverAttempted)
+            | (Failed, Failed)
+            | (AttemptPending, AttemptPending) => Ordering::Equal,
+            // We reconnect to `Responded` peers that have stopped sending messages,
+            // then `NeverAttempted` peers, then `Failed` peers
+            (Responded, _) => Ordering::Less,
+            (_, Responded) => Ordering::Greater,
+            (NeverAttempted, _) => Ordering::Less,
+            (_, NeverAttempted) => Ordering::Greater,
+            (Failed, _) => Ordering::Less,
+            (_, Failed) => Ordering::Greater,
+            // AttemptPending is covered by the other cases
+        }
+    }
+}
+
+impl PartialOrd for PeerConnectionState {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
```

Comment:
Should this be defined in the `peer` module and used in here for `MetaAddr`? If yes we should reconcile this type of `peer::connection::State` with [the existing one](https://github.com/ZcashFoundation/zebra/blob/cc7d5bd2adea1fee9c9092f0438c5b04d5cf91b3/zebra-network/src/peer/connection.rs#L311), which is an enum that only has variants about `awaiting` or `failed`.

Reply:
No, these are very different states. peer::connection::State tracks the state of a single TCP connection. This state tracks peer address activity.

So I think we should rename this state to PeerAddrState.

Reply:
I like that 👍